### PR TITLE
Loosen property name collision detection involving hidden properties

### PR DESCRIPTION
**Pull request by layomia**
The change in https://github.com/dotnet/runtime/pull/32107 allowed the (de)serialization of hidden properties when the properties in the derived classes are ignored/non-public.

As a result of the change, the serializer now examines properties at all levels of the inheritance chain, rather than just the ones visible from the type to be serialized. This means that `InvalidOperationException` could be thrown if there is a JSON property name conflict between inheritance levels, if the property higher in the inheritance chain was not hidden by the conflicting property in the more derived class (but by another property). A property is "hidden" if it is virtual and overridden in a derived class with polymorphism or with the `new` keyword.

The breaking change was flagged in ASP.NET Core preview 5 validation: dotnet/aspnetcore#22132.

To fix the break, this change loosens property name collision detection involving hidden properties.  Rather than throw `InvalidOperationException`, the serializer will a ignore hidden property when the property that hid it is ignored _and_ there's a property name conflict. This aligns with Newtonsoft.Json behavior.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 1]
```diff
@@ -101,6 +101,9 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                                 ? StringComparer.OrdinalIgnoreCase
                                 : StringComparer.Ordinal);
 
+                        HashSet<string> ignoredProperties = new HashSet<string>();
```
Comment (BrennanConroy) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:104]: Would avoiding this allocation until needed and adding null checks be better?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 2]
```diff
@@ -111,31 +114,43 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
+                                // For now we only support public properties (i.e. setter and/or getter is public).
                                 if (propertyInfo.GetMethod?.IsPublic == true ||
                                     propertyInfo.SetMethod?.IsPublic == true)
                                 {
                                     JsonPropertyInfo jsonPropertyInfo = AddProperty(propertyInfo, currentType, options);
                                     Debug.Assert(jsonPropertyInfo != null && jsonPropertyInfo.NameAsString != null);
 
-                                    // If the JsonPropertyNameAttribute or naming policy results in collisions, throw an exception.
+                                    string propertyName = propertyInfo.Name;
+                                    bool isIgnored = jsonPropertyInfo.IsIgnored;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:125]: Seems like this temp variable can be removed since it is only used once. Doing that would also make it a bit more readable.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 3]
```diff
@@ -111,31 +114,43 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
+                                            // Overwrite previously cached property since it has [JsonIgnore].
                                             cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
                                         }
-                                        else if (other.PropertyInfo?.Name != jsonPropertyInfo.PropertyInfo?.Name &&
-                                            (jsonPropertyInfo.ShouldDeserialize == true || jsonPropertyInfo.ShouldSerialize == true))
+                                        else if (!(isIgnored || other.PropertyInfo!.Name == propertyName || ignoredProperties?.Contains(propertyName) == true))
                                         {
-                                            // Check for name equality is required to determine when a new slot is used for the member.
-                                            // Therefore, if names are not the same, there is conflict due to the name policy or attributes.
+                                            // The collision is invalid if none of the following is true:
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:139]: This is hard to follow. Can we just state the conditions when we do throw?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 4]
```diff
@@ -111,31 +114,43 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
-                                        if (other.ShouldDeserialize == false && other.ShouldSerialize == false)
+                                        if (other.IsIgnored)
                                         {
-                                            // Overwrite the one just added since it has [JsonIgnore].
+                                            // Overwrite previously cached property since it has [JsonIgnore].
                                             cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
                                         }
-                                        else if (other.PropertyInfo?.Name != jsonPropertyInfo.PropertyInfo?.Name &&
-                                            (jsonPropertyInfo.ShouldDeserialize == true || jsonPropertyInfo.ShouldSerialize == true))
+                                        else if (!(isIgnored || other.PropertyInfo!.Name == propertyName || ignoredProperties?.Contains(propertyName) == true))
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:137]: FWIW I find it's more readable to avoid having to translate the ! operator against each condition:
```
else if (!isIgnored &&
    other.PropertyInfo!.Name != propertyName &&
    !ignoredProperties?.Contains(propertyName)))
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 5]
```diff
@@ -111,31 +114,43 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
+                                            //    so we'll ignore this one as well.
                                             ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, jsonPropertyInfo);
                                         }
-                                        // else ignore jsonPropertyInfo since it has [JsonIgnore] or it's hidden by a new slot.
+                                        // Ignore the current property.
+                                    }
+
+                                    if (jsonPropertyInfo.IsIgnored)
+                                    {
+                                        (ignoredProperties ??= new HashSet<string>()).Add(propertyName);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:153]: Does this have to respect the case-insensitive mode?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 4 above)
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:137]: is `other.PropertyInfo!.Name == propertyName` protecting against duplicate properties via `[JsonPropertyName]` and\or duplicate properties by casing (when using case-insensitive options?). Please add a comment around this.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:137]: `other.PropertyInfo!.Name == propertyName` indicates that the previously cached property ("other") hides the current property (either with the `new` keyword, or by overriding it). We shouldn't throw in this case, as it is likely for the two properties result in the same property name (since they have the same CLR name). Rather, we should ignore the current property.

I've added a comment.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:139]: I've simplified the comments.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:153]: This cache tells us the CLR names for properties that have been ignored. This is used when looking at properties of a base type, and there's a conflict with a property at a derived type (which has a different CLR name, but the same JSON property name, due to `JsonPropertyName` or the naming policy).

If another property on a derived class has the same CLR name as the current property on the base type, we know that the former hid the latter. If the property on the derived class was ignored with `JsonIgnore`, we know to also ignore the current type being examined on the base type. This will mitigate the property name collision.

None of these considerations has to do with case-insensitivity.

**Review by steveharter — APPROVED** [review]
Per offline review\discussion, these changes look good.

Notes:
- [JsonPropertyName] does not impact "hiding". A base property that has the attribute (with name "foo") will not hide a more derived type with a "foo" CLR property -- instead it throws InvalidOperation. The same applies if the derived type has [JsonPropertyName] and a base property also has a "foo" CLR property.
- A derived type hides a base type property of the same name (return type doesn't matter). The name can be case-insensitive if the JsonSerializerOptions are set for that.
